python/test/executor_test.py

```diff
@@ -33,12 +48,58 @@ def bench_time(niters: int, ngraphIters: int, func):
     end = cp.cuda.Event()
 
     start.record(stream)
-    for _ in range(ngraphIters):
+    for _ in range(n_graph_iters):
         graph.launch(stream)
     end.record(stream)
     end.synchronize()
 
-    return cp.cuda.get_elapsed_time(start, end) / niters * 1000.0 / ngraphIters
+    return cp.cuda.get_elapsed_time(start, end) / n_iters * 1000.0 / n_graph_iters
+
+
+def bench_correctness(
+    execution_plan_name: str,
+    input_buf: cp.ndarray,
+    result_buf: cp.ndarray,
+    test_buf: cp.ndarray,
+    dtype_str: str,
+    rank: int,
+    num_ranks: int,
+    n_iters: int,
+    func,
+):
+    type_size = cp.dtype(parse_dtype(dtype_str)).itemsize
+
+    fill_data_kernel_name = "fill_data_%s" % dtype_str
+    if "allgather" in execution_plan_name:
+        coll = "all_gather"
+    else:
+        coll = "all_reduce"
+    test_data_kernel_name = "test_data_%s_%s" % (coll, dtype_str)
+
+    file_dir = os.path.dirname(os.path.abspath(__file__))
+    fill_data_kernel = KernelBuilder(
+        file="executor_test_verifier.cu", kernel_name=fill_data_kernel_name, file_dir=file_dir
+    ).get_compiled_kernel()
+    test_data_kernel = KernelBuilder(
+        file="executor_test_verifier.cu", kernel_name=test_data_kernel_name, file_dir=file_dir
+    ).get_compiled_kernel()
+    nblocks = 64
+    nthreads = 1024
+
+    stream = cp.cuda.Stream(non_blocking=True)
+    with stream:
+        stream.begin_capture()
+        for i in range(n_iters):
+            fill_data_params = pack(input_buf) + struct.pack("Q", input_buf.nbytes // type_size) + pack(rank, i)
```

Comment:
A bit confused, why we don't use function from cupy to init data and verify the correctness but write the kernel. Because of the limitation of cuda graph?

Reply:
cupy calls need extra CUDA synchronization to know whether the results are correct at CPU side. I want to put the check inside the kernel to minimize such synchronization, to better make sure there is no barrier issue or data corruption issue in asynchronous kernel execution conditions. It’s actually more like running a real GPU application.